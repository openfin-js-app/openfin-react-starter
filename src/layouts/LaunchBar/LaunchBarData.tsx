import AccessibilityIcon from '@material-ui/icons/Accessibility';
import AccountBoxIcon from '@material-ui/icons/AccountBox';
import BookmarkIcon from '@material-ui/icons/Bookmark';
import BackupIcon from '@material-ui/icons/Backup';
import BugReportIcon from '@material-ui/icons/BugReport';
import LanguageIcon from '@material-ui/icons/Language';
import ReportIcon from '@material-ui/icons/Report';

import svgNumberOne from'../../assets/svg/other/number-1.svg';
import svgNumberTwo from'../../assets/svg/other/number-2.svg';
import googleSearch from'../../assets/svg/other/google-search.svg';

export interface IlaunchBarItemType {
    icon:any,
    disabled:boolean,
    svg:string,
    appJson:any,
}

export const sampleItems:IlaunchBarItemType[]=[
    {
        icon:AccessibilityIcon,
        disabled:false,
        svg:null,
        appJson:{
            name:'openfin-react-starter-child-accessibility',
            url:'/childWindow/accessibility',
            frame:false,
            resizable:true,
            state:'normal',
            autoShow:true,
            minWidth:520,
            minHeight:170,
        },
    },
    {
        icon:null,
        disabled:false,
        svg:svgNumberOne,
        appJson:{
            name:'openfin-react-starter-child-view-one',
            url:'/childWindow/view-one',
            frame:false,
            resizable:true,
            state:'normal',
            autoShow:true,
            minWidth:415,
            minHeight:100,
        },
    },
    {
        icon:null,
        disabled:false,
        svg:svgNumberTwo,
        appJson:{
            name:'openfin-react-starter-child-view-two',
            url:'/childWindow/view-two',
            frame:false,
            resizable:true,
            state:'normal',
            autoShow:true,
            minWidth:415,
            minHeight:100,
        },
    },
    {
        icon:ReportIcon,
        disabled:false,
        svg:null,
        appJson:{
            name:'openfin-react-starter-child-report',
            url:'/childWindow/report',
            frame:false,
            resizable:true,
            state:'normal',
            autoShow:true,
            minWidth:430,
            minHeight:100,
        },
    },
    {
        icon:null,
        disabled:false,
        svg:googleSearch,
        appJson:{
            name:'openfin-react-starter-child-google',
            url:'https://www.google.com/',
            frame:true,
            resizable:true,
            state:'normal',
            autoShow:true,
            callback:(window:any)=>{
                console.log('onOpenNewSelf::callback', window);
            }
        },
    },
];

export const disabledItems:IlaunchBarItemType[]=[
    {
        icon:AccountBoxIcon,
        disabled:true,
        svg:null,
        appJson:{},
    },
    {
        icon:BookmarkIcon,
        disabled:true,
        svg:null,
        appJson:{},
    },
    {
        icon:BackupIcon,
        disabled:true,
        svg:null,
        appJson:{},
    },
    {
        icon:BugReportIcon,
        disabled:true,
        svg:null,
        appJson:{},
    },
    {
        icon:LanguageIcon,
        disabled:true,
        svg:null,
        appJson:{},
    },
];

export const launchBarItems = [
    ...sampleItems,
    ...disabledItems,
];