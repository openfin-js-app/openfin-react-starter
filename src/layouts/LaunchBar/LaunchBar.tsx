import * as React from 'react';
import { connect } from 'react-redux';
import cx from 'classnames';
import { Scrollbars } from 'react-custom-scrollbars';

import { WithStyles, withStyles, StyleRules } from '@material-ui/core/styles';

import { Window } from '@albertli90/redux-openfin';

import AppBar from '@material-ui/core/AppBar';
import Toolbar from '@material-ui/core/Toolbar';
import Button from '@material-ui/core/Button';
import IconButton from '@material-ui/core/IconButton';

import ArrowBackIcon from '@material-ui/icons/ArrowBack';
import ArrowForwardIcon from '@material-ui/icons/ArrowForward';
import ClearIcon from '@material-ui/icons/Clear';
import RemoveIcon from '@material-ui/icons/Remove';
import ZoomOutMapIcon from '@material-ui/icons/ZoomOutMap';

import { launchBarLayoutStyle as style } from '../../assets/jss/openfin-starter';

import appLogo from '../../assets/svg/app.svg';

import {
    // actions
    applicationLaunchBarToggle, applicationLaunchBarToggleCollapse, applicationLaunchNewWindow,
    // types
    IRootState,
} from '../../redux';

import { launchBarItems } from './LaunchBarData';

interface IProps extends WithStyles<typeof style>{
    launchBarCollapse:boolean,
    actions:{
        handleLaunchBarItemBtnClick: (appJson:any)=>()=>void,
        handleSwitchToMainWindow: ()=>void,
        handleToggleCollapse: ()=>void,
        handleMinimize: ()=>void,
        handleClose: ()=>void,
    }
}

class LaunchBarComp extends React.Component<IProps,{}>{
    render(){

        const {
            classes, launchBarCollapse,
            actions:{
                handleLaunchBarItemBtnClick,
                handleToggleCollapse,
                handleSwitchToMainWindow,
                handleMinimize,
                handleClose,
            }
        } = this.props;

        const collapse = launchBarCollapse;

        // const _launchBarItems = launchBarItems;

        const buttonContainerWidth = launchBarItems.length<10?launchBarItems.length*64:576;

        return (
            <span>
                <AppBar position={"static"}>
                    <Toolbar className={classes.toolBar}>
                        <img src={appLogo} className={classes.appLogoImg}/>
                        <Scrollbars
                            className={cx(
                                classes.buttonContainer,
                                {
                                    [classes.buttonContainerCollapse]:collapse,
                                }
                            )}
                            style={{
                                width:`${buttonContainerWidth}px`,
                            }}
                        >
                            {launchBarItems.map((item,index)=>{
                                if (item.icon){
                                    return <IconButton key={index}
                                       className={classes.baseButton}
                                       disabled={item.disabled}
                                       onClick={handleLaunchBarItemBtnClick(item.appJson)}
                                    >
                                            {React.createElement(item.icon)}
                                    </IconButton>
                                }else{
                                    return <IconButton key={index}
                                       className={classes.svgButton}
                                       disabled={item.disabled}
                                       onClick={handleLaunchBarItemBtnClick(item.appJson)}
                                    >
                                        <img src={item.svg}/>
                                    </IconButton>
                                }
                            })}
                        </Scrollbars>
                        <div className={classes.controlPanelContainer}>
                            <Button
                                className={cx(classes.controlBtn, classes.controlBtnInfo)}
                                variant={'fab'} mini color={'secondary'} aria-label={'collapse'}
                                onClick={handleToggleCollapse}
                            >
                                {collapse?
                                    <ArrowForwardIcon/>:
                                    <ArrowBackIcon/>
                                }
                            </Button>
                            <Button
                                className={cx(classes.controlBtn, classes.controlBtnSuccess)}
                                variant={'fab'} mini color={'secondary'} aria-label={'minimize'}
                                onClick={handleMinimize}
                            >
                                <RemoveIcon/>
                            </Button>
                            <Button
                                className={cx(classes.controlBtn, classes.controlBtnWarning)}
                                variant={'fab'} mini color={'secondary'} aria-label={'expand'}
                                onClick={handleSwitchToMainWindow}
                            >
                                <ZoomOutMapIcon/>
                            </Button>
                            <Button
                                className={cx(classes.controlBtn, classes.controlBtnDanger)}
                                variant={'fab'} mini color={'secondary'} aria-label={'close'}
                                onClick={handleClose}
                            >
                                <ClearIcon/>
                            </Button>
                        </div>
                    </Toolbar>
                </AppBar>
            </span>
        );
    }
}

export default connect(
    (state:IRootState)=>({
        launchBarCollapse:state.application.launchBarCollapse,
    }),
    dispatch => ({
        actions:{
            handleLaunchBarItemBtnClick:(appJson)=>()=>{dispatch(applicationLaunchNewWindow(appJson))},
            handleSwitchToMainWindow:()=>{dispatch(applicationLaunchBarToggle())},
            handleToggleCollapse:()=>{dispatch(applicationLaunchBarToggleCollapse())},
            handleMinimize:()=>{dispatch(Window.actions.minimize({}))},
            handleClose:()=>{dispatch(Window.actions.close({force:false}))},
        }
    })
)(withStyles(style)(LaunchBarComp));