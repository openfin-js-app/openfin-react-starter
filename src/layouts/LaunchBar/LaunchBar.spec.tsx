import * as React from 'react';
import { Provider } from 'react-redux';
import { createMuiTheme } from '@material-ui/core/styles';
import { ThemeProvider } from '@material-ui/styles';
import { createShallow, createMount } from '@material-ui/core/test-utils';
import configurestore from 'redux-mock-store';

import Button from '@material-ui/core/Button';
import IconButton from '@material-ui/core/IconButton';

import LaunchBar from './LaunchBar';

const muiTheme = createMuiTheme({});

const mockStore = configurestore();
const initialState = {
    application:{
        launchBarCollapse:false,
    }
};

describe('LaunchBar Layout',()=>{

    let shallow;
    let mount;

    beforeAll(() => {
        mount = createMount();
    });

    afterAll(() => {
        mount.cleanUp();
    });


    beforeEach(() => {
        shallow = createShallow();
    });

    it('render correctly and launch all non-disabled windows',()=>{
        const store = mockStore(initialState);
        const wrapper = mount(
            <ThemeProvider theme={muiTheme}>
                <Provider store={store}>
                    <LaunchBar/>
                </Provider>
            </ThemeProvider>
        );
        wrapper.find(IconButton).forEach((iconBtn)=>{
            const props = iconBtn.props();
            if ( props.disabled === false && typeof props.onClick === 'function'){
                props.onClick();
            }
        });
        expect(store.getActions()).toMatchSnapshot();
    });

    it('trigger all control panel btns',()=>{
        const store = mockStore(initialState);
        const wrapper = mount(
            <ThemeProvider theme={muiTheme}>
                <Provider store={store}>
                    <LaunchBar/>
                </Provider>
            </ThemeProvider>
        );
        wrapper.find(Button).forEach((btn)=>{
            const props = btn.props();
            if ( typeof props.onClick === 'function'){
                props.onClick();
            }
        });
        expect(store.getActions()).toMatchSnapshot();
    });

    it('render collapsed button container',()=>{
        const store = mockStore({
            application:{
                launchBarCollapse:true,
            }
        });
        const wrapper = mount(
            <ThemeProvider theme={muiTheme}>
                <Provider store={store}>
                    <LaunchBar/>
                </Provider>
            </ThemeProvider>
        );
        expect(wrapper).toBeTruthy();
    });

});