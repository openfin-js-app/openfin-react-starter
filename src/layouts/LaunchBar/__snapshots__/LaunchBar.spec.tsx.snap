// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LaunchBar Layout render correctly and launch all non-disabled windows 1`] = `
Array [
  Object {
    "payload": Object {
      "autoShow": true,
      "frame": false,
      "minHeight": 170,
      "minWidth": 520,
      "name": "openfin-react-starter-child-accessibility",
      "resizable": true,
      "state": "normal",
      "url": "/childWindow/accessibility",
    },
    "type": "APPLICATION_LAUNCH_NEW_WINDOW",
  },
  Object {
    "payload": Object {
      "autoShow": true,
      "frame": false,
      "minHeight": 100,
      "minWidth": 415,
      "name": "openfin-react-starter-child-view-one",
      "resizable": true,
      "state": "normal",
      "url": "/childWindow/view-one",
    },
    "type": "APPLICATION_LAUNCH_NEW_WINDOW",
  },
  Object {
    "payload": Object {
      "autoShow": true,
      "frame": false,
      "minHeight": 100,
      "minWidth": 415,
      "name": "openfin-react-starter-child-view-two",
      "resizable": true,
      "state": "normal",
      "url": "/childWindow/view-two",
    },
    "type": "APPLICATION_LAUNCH_NEW_WINDOW",
  },
  Object {
    "payload": Object {
      "autoShow": true,
      "frame": false,
      "minHeight": 100,
      "minWidth": 430,
      "name": "openfin-react-starter-child-report",
      "resizable": true,
      "state": "normal",
      "url": "/childWindow/report",
    },
    "type": "APPLICATION_LAUNCH_NEW_WINDOW",
  },
  Object {
    "payload": Object {
      "autoShow": true,
      "callback": [Function],
      "frame": true,
      "name": "openfin-react-starter-child-google",
      "resizable": true,
      "state": "normal",
      "url": "https://www.google.com/",
    },
    "type": "APPLICATION_LAUNCH_NEW_WINDOW",
  },
]
`;

exports[`LaunchBar Layout trigger all control panel btns 1`] = `Array []`;
